src/insert/spacefinder/spacefinder.ts

```diff
@@ -198,22 +203,79 @@ const partitionCandidates = <T>(
 	return [filtered, exclusions];
 };
 
+/**
+ * Check if the top of the candidate is far enough from the opponent
+ *
+ *                                                        │
+ *                     Candidate Below                    │             Candidate Above
+ *                                                        │
+ *                  ───────────────────  Top of container │          ───────────────────  Top of container
+ *                    ▲              ▲                    │            ▲              ▲
+ *                    │              │                    │            │              │ candidate.top
+ *                    │ ┌──────────┐ │                    │            │ ┌──────────┐ ▼   (insertion point)
+ *                    │ │          │ │opponent.bottom     │            │ │          │
+ *                    │ │ Opponent │ │                    │            │ │ Candidate│
+ *       candidate.top│ │          │ │                    │opponent.top│ │          │
+ *                    │ └──────────┘ ▼                    │            │ └──────────┘
+ *                    │           ▲                       │            │
+ *                    │           │ minBelow              │            │  ───────────
+ *                    │           ▼                       │            │           ▲
+ *                    │ ────────────                      │            │           │ minAbove
+ *                    │                                   │            │           ▼
+ * (insertion point)  ▼ ┌──────────┐                      │            ▼ ┌──────────┐
+ *                      │          │                      │              │          │
+ *                      │ Candidate│                      │              │ Opponent │
+ *                      │          │                      │              │          │
+ *                      └──────────┘                      │              └──────────┘
+ *                                                        │
+ *                                                        │
+ *
+ * @param candidate
+ * @param opponent
+ * @param rule
+ * @param isCandidateBelow
+ * @returns
+ */
+const isTopOfCandidateFarEnoughFromOpponent = (
+	candidate: SpacefinderItem,
+	opponent: SpacefinderItem,
```

Comment:
Is the opponent a previous ad slot or another element\candidate?

Reply:
Opponents are other elements in the article that are in the spacefinder ruleset for the current pass. This includes slots inserted by a previous pass but not those in the current pass as they're all inserted at the end.

Reply:
Are those in the spacefinder ruleset defined by `slotSelector`? For example, on mobile the opponents are other elements that are `p`, `h2` or `hr`?

Reply:
`slotSelector` is for valid candidates, the `selectors` is for opponents to avoid.